loadBalancer/src/test/scala/com/github/pandafolks/panda/loadbalancer/ConsistentHashingStateTest.scala

```diff
@@ -245,4 +247,33 @@ class ConsistentHashingStateTest extends AsyncFlatSpec with PrivateMethodTester
         underTest.invokePrivate(usedIdentifiersWithPositions()).keySet().asScala.count(k => k.identifier.startsWith(group3.name)) should be (400)
     }
   }
+
+  it should "clear empty groups" in {
+    val underTest = new ConsistentHashingState(positionsPerIdentifier = 500)
+    val usedPositionsGroupedByGroup = PrivateMethod[ConcurrentHashMap[Group, TreeMap[Int, Participant]]](Symbol("usedPositionsGroupedByGroup"))
+
+    val group1 = Group("cars")
+    val group2 = Group("planes")
+
+    for (i <- 1 to 500) {
+      underTest.add(Participant("ip", 123, group1, group1.name + i))
+      underTest.add(Participant("ip", 123, group2, group2.name + i))
+    }
+
+    var futures: List[Future[Unit]] = List()
+    for (i <- 1 to 500) {
+      futures = Future(underTest.remove(Participant("ip", 123, group1, group1.name + i))) :: futures
+      futures = Future(underTest.remove(Participant("ip", 123, group2, group2.name + i))) :: futures
+    }
+    val futureWithList: Future[List[Unit]] = futures.sequence
+
+    whenReady(futureWithList, Timeout.apply(Span.apply(10, Seconds))) {
+      _ =>
+        underTest.invokePrivate(usedPositionsGroupedByGroup()).keySet().size() should be(2)
+        underTest.invokePrivate(usedPositionsGroupedByGroup()).get(group1).isEmpty should be(true)
+        underTest.invokePrivate(usedPositionsGroupedByGroup()).get(group2).isEmpty should be(true)
+        Await.result(underTest.clearEmptyGroups().runToFuture, 5.seconds)
```

Comment:
Generally, it's very bad to put Await.result inside `whenReady`. In this piece of code, it is possible to get all results of `usedPositionsGroupedByGroup` invocation earlier (inside `future` pipeline ofc) and calling `clearEmptyGroups()` as a last part of the `future` pipeline. However, I'm ok with this solution and you can leave it as it is. Just worth remembering in the future.